Handle deployment to older camunda versions that don't expose their version.

michaelarnauts opened the pull request:
This PR adds a try catch around version retrieval and only send it in the event when we have a version.

This fixes a regression in 4.8.0 that broke deployment to some older camunda versions (7.8, but newer might also be impacted).

See #2340 for more info.

@barmac

barmac commented on client/src/plugins/camunda-plugin/deployment-tool/DeploymentTool.js:
```diff
@@ -135,7 +135,11 @@ export default class DeploymentTool extends PureComponent {
 
       // (3.1) Retrieve version we deploy to via API
-      version = (await this.getVersion(configuration)).version;
+      try {
+        version = (await this.getVersion(configuration)).version;
+      } catch (error) {
+        version = null;
+      }
```
This makes the function swallow all errors. I'd rather like to avoid a situation when there's a bug and we don't know about it because no errors are thrown. Can you please make sure that we ignore the version only in case of 404 errors?

nikku replied:
Also: We're missing test coverage here.

barmac reviewed (changes requested):
Hi,

Thank you for your contribution. Before we merge this, please:

* Make sure we only ignore 404 errors (cf. the comment on the code).
* Make sure the commit message adheres to [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) rules. Note that the commit body should end with "Closes #2340".
* Add test cases for the code you added.

If you have any questions, feel free to tag me.

nikku commented:
@michaelarnauts Do you plan to follow up on your PR to incorporate the review feedback we provided?

nikku commented on client/src/plugins/camunda-plugin/deployment-tool/DeploymentTool.js:
```diff
@@ -167,16 +171,17 @@ export default class DeploymentTool extends PureComponent {
-        },
-        context: 'deploymentTool'
+        context: 'deploymentTool',
+        ...(deployedTo && { deployedTo: deployedTo })
```
Missing test coverage here, too.

I think we can safely set `deployedTo` to `null` to indicate we could have set it, but did not retrieve any.

michaelarnauts replied:
This was already in the codebase for `handleDeploymentError`. It omitted the `deployedTo` when there was no `deployedTo` available.

nikku commented on client/src/plugins/camunda-plugin/start-instance-tool/StartInstanceTool.js:
```diff
@@ -222,7 +222,11 @@ export default class StartInstanceTool extends PureComponent {
       // (5.1) Retrieve version via API
-      version = (await this.getVersion(deploymentConfig)).version;
+      try {
+        version = (await this.getVersion(deploymentConfig)).version;
```
Missing test coverage for this behavior.

michaelarnauts commented:
> @michaelarnauts Do you plan to follow up on your PR to incorporate the review feedback we provided?

Sure, I'll try to look into it this week.

michaelarnauts commented:
I've modified the `getVersion` function now to return null. The exception we got wasn't a 404, but a syntax error since it was trying to parse an empty string as json.

I've added a test and ran it with `npm run app:test`, but I'm not sure if the test are actually ran locally. I assume I still have to remove the `version: ` from the `getVersionSpy`'s.

michaelarnauts commented on client/src/plugins/camunda-plugin/deployment-tool/__tests__/DeploymentToolSpec.js:
```diff
@@ -498,6 +498,25 @@ describe('<DeploymentTool>', () => {
+
+      // then
+      expect(getVersionSpy).to.have.been.calledOnce;
+      expect(getVersionSpy).to.not.have.been.called;
```
I expect this to at throw an error, but it doesn't. I'm not sure if this test is ran.

michaelarnauts replied:
Never mind, it was `npm run client:test` that was needed to run the tests.

michaelarnauts commented:
I'm not sure what the correct behaviour will be. Should the getVersion return `null`, or should it return `{ version: null }`? There seems to be a test that expect it to be `undefined` (should return undefined if version cannot be parsed), but that's not how the code behaves, since there is no check if version is actually undefined.

michaelarnauts commented:
This is the cleanest I have it now. All tests pass. 

I don't know how to properly throw an `ConnectionError` in the mocked `getVersion` however, so the test isn't really testing that ConnectionError.

barmac commented:
Merged via https://github.com/camunda/camunda-modeler/commit/a89ef727fcba8c21c522ca3d37523ff64d0d31a7

barmac commented:
Thank you for your contribution! 🎉

nikku commented:
Very nice contribution indeed :clap:
